internal/controller/postgrescluster/apply_test.go

```diff
@@ -159,48 +159,24 @@ func TestServerSideApply(t *testing.T) {
 		}
 
 		reconciler := Reconciler{Client: cc, Owner: client.FieldOwner(t.Name())}
+		upstream := constructor("status-upstream")
 
-		// Start with fields filled out.
-		intent := constructor("change-to-zero")
-		intent.Spec.Template.Spec.SecurityContext = &corev1.PodSecurityContext{
-			SupplementalGroups: []int64{1, 2, 3},
-		}
-
-		// Create the StatefulSet.
-		before := intent.DeepCopy()
-		assert.NilError(t,
-			cc.Patch(ctx, before, client.Apply, client.ForceOwnership, reconciler.Owner))
-
-		// Change fields to zero.
-		intent.Spec.Template.Spec.SecurityContext.SupplementalGroups = nil
-
-		// client.Apply cannot correct it in old versions of Kubernetes.
-		after := intent.DeepCopy()
-		assert.NilError(t,
-			cc.Patch(ctx, after, client.Apply, client.ForceOwnership, reconciler.Owner))
-
+		// The structs defined in "k8s.io/api/apps/v1" marshal empty status fields.
 		switch {
-		case serverVersion.LessThan(version.MustParseGeneric("1.18.19")):
-
-			// - https://pr.k8s.io/101179
-			assert.Assert(t, !equality.Semantic.DeepEqual(
-				after.Spec.Template.Spec.SecurityContext,
-				intent.Spec.Template.Spec.SecurityContext),
-				"expected https://issue.k8s.io/89273, got %v",
-				after.Spec.Template.Spec.SecurityContext)
+		case serverVersion.LessThan(version.MustParseGeneric("1.22")):
+			assert.ErrorContains(t,
+				cc.Patch(ctx, upstream, client.Apply, client.ForceOwnership, reconciler.Owner),
+				"field not declared in schema",
+				"expected https://issue.k8s.io/109210")
 
 		default:
-			assert.DeepEqual(t,
-				after.Spec.Template.Spec.SecurityContext,
-				intent.Spec.Template.Spec.SecurityContext)
+			assert.NilError(t,
+				cc.Patch(ctx, upstream, client.Apply, client.ForceOwnership, reconciler.Owner))
 		}
 
-		// Our apply method corrects it.
-		again := intent.DeepCopy()
+		// Our apply method generates the correct apply-patch.
+		again := constructor("status-local")
 		assert.NilError(t, reconciler.apply(ctx, again))
```

Comment:
Just to walk through this,
(a) we try to client.Patch, which fails below 1.22, and succeeds above that version
(b) we try to use our `apply`, which with a statefulset now passes it through to reconciler.patch which passes it to client.Patch

But we're not setting/checking the SecurityContext in either case (because in both cases it will be there)?

Reply:
> Just to walk through this,
> (a) we try to client.Patch, which fails below 1.22, and succeeds above that version
> (b) we try to use our `apply`, which with a statefulset now passes it through to reconciler.patch which passes it to client.Patch

Right. `client.Patch` marshals the `StatefulSet` struct which lacks an `omitempty` and fails on certain server versions. Our method does more than one marshal to produce an apply-patch without that field.

> But we're not setting/checking the SecurityContext in either case (because in both cases it will be there)?

The diff makes it look like one test changed, but it's really a new test. This isn't about the security context anymore.

Reply:
`upstream`, `local`, `again`... If there are comments or names that can be improved to make this clear, go for it!